src/java.base/share/classes/jdk/internal/reflect/ReflectionFactory.java

```diff
@@ -623,74 +587,135 @@ public final boolean hasStaticInitializerForSerialization(Class<?> cl) {
+        return (config().useDirectMethodHandle & Config.METHOD_MH_ACCESSOR) == Config.METHOD_MH_ACCESSOR;
     }
 
     static boolean useFieldHandleAccessor() {
-        return (useDirectMethodHandle & FIELD_MH_ACCESSOR) == FIELD_MH_ACCESSOR;
+        return (config().useDirectMethodHandle & Config.FIELD_MH_ACCESSOR) == Config.FIELD_MH_ACCESSOR;
     }
 
     static boolean useNativeAccessorOnly() {
-        return useNativeAccessorOnly;
+        return config().useNativeAccessorOnly;
+    }
+
+    private static boolean disableSerialConstructorChecks() {
+        return config().disableSerialConstructorChecks;
     }
 
     /** We have to defer full initialization of this class until after
         the static initializer is run since java.lang.reflect.Method's
         static initializer (more properly, that for
         java.lang.reflect.AccessibleObject) causes this class's to be
         run, before the system properties are set up. */
-    private static void checkInitted() {
-        if (initted) return;
+    private static Config config;
 
+    private static Config config() {
         // Defer initialization until module system is initialized so as
         // to avoid inflation and spinning bytecode in unnamed modules
         // during early startup.
         if (!VM.isModuleSystemInited()) {
-            return;
+            return Config.fallback;
         }
 
-        Properties props = GetPropertyAction.privilegedGetProperties();
-        String val = props.getProperty("sun.reflect.noInflation");
-        if (val != null && val.equals("true")) {
-            noInflation = true;
+        Config c = config;
+        if (c == null) {
+            config = c = new Config(true);
         }
+        return c;
+    }
 
-        val = props.getProperty("sun.reflect.inflationThreshold");
-        if (val != null) {
-            try {
-                inflationThreshold = Integer.parseInt(val);
-            } catch (NumberFormatException e) {
-                throw new RuntimeException("Unable to parse property sun.reflect.inflationThreshold", e);
-            }
-        }
-        val = props.getProperty("jdk.reflect.useDirectMethodHandle");
-        if (val != null) {
-            if (val.equals("false")) {
-                useDirectMethodHandle = 0;
-            } else if (val.equals("methods")) {
-                useDirectMethodHandle = METHOD_MH_ACCESSOR;
-            } else if (val.equals("fields")) {
-                useDirectMethodHandle = FIELD_MH_ACCESSOR;
-            }
-        }
-        val = props.getProperty("jdk.reflect.useNativeAccessorOnly");
-        if (val != null && val.equals("true")) {
-            useNativeAccessorOnly = true;
-        }
+    /**
+     * The configurations exist as an object to avoid race conditions.
+     * See bug 8261407.
+     */
+    private static final class Config {
+        private static final Config fallback = new Config(false);
```

Comment:
```suggestion
        private static final Config DEFAULT = new Config(false);
```

I suggest to rename this to `DEFAULT`.